sda/cmd/rotatekey/rotatekey.go

```diff
@@ -0,0 +1,268 @@
+
+				continue
+			}
+			if newHeader == nil {
+				err := errors.New("reencrypt returned empty header")
+				msg := fmt.Sprintf("failed to rotate c4gh key for file %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			// Rotate header and keyhash in database
+			if err := db.RotateHeaderKey(newHeader, keyhash, fileID); err != nil {
+				msg := fmt.Sprintf("RotateHeaderKey failed for file-id: %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			aID, err := db.GetAccessionID(fileID)
+			if err != nil {
+				msg := fmt.Sprintf("GetAccessionID failed for file-id: %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			// Send re-verify message
+			reVerify, err := db.GetReVerificationData(aID)
+			if err != nil {
+				msg := fmt.Sprintf("GetReVerificationData failed for file-id %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			reVerifyMsg, _ := json.Marshal(&reVerify)
+			err = schema.ValidateJSON(fmt.Sprintf("%s/ingestion-verification.json", conf.Broker.SchemasPath), reVerifyMsg)
+			if err != nil {
+				msg := "Validation of outgoing re-verify message failed"
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			if err := mq.SendMessage(delivered.CorrelationId, conf.Broker.Exchange, "archived", reVerifyMsg); err != nil {
+				msg := "failed to publish message"
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			if err := delivered.Ack(false); err != nil {
+				log.Errorf("failed to Ack message, reason: (%s)", err.Error())
+			}
+		}
+	}()
+
+	<-forever
+}
+
+// Nack message and send the payload to an error queue so it can be analyzed.
+func nackAndSendToErrorQueueckAndSendToErrorQueue(mq *broker.AMQPBroker, delivered amqp091.Delivery, mqExchange, msg, reason string) {
+	infoErrorMessage := broker.InfoError{
+		Error:           msg,
+		Reason:          reason,
+		OriginalMessage: string(delivered.Body),
+	}
+	body, _ := json.Marshal(infoErrorMessage)
+
+	if err := mq.SendMessage(delivered.CorrelationId, mqExchange, "error", body); err != nil {
+		log.Errorf("failed to publish message, reason: (%s)", err.Error())
+	}
+	if err := delivered.Nack(false, false); err != nil {
+		log.Errorf("failed to Ack message, reason: (%s)", err.Error())
+	}
+}
```

Comment:
Just because you get an error it doesn't mean that the message should be `nacked`. Look at the logic from within the MQTT and you will see how the different cases work.

* `nack/reject with requeue=1`: the message will be returned to the queue it came from as though it were a new message; this might be useful in case of a temporary failure on the consumer side
* `nack/reject with requeue=0 and a configured Dead Letter Exchange (DLX)`:  publish the message to that exchange, allowing it to be picked up by another queue
* `nack/reject with requeue=0 and no DLX`: simply discard the message
* `ack` will remove the message from the queue even if a DLX is configured

Reply:
That logic is flawed. The ack/nack is about how the MQTT should deal with the message not if the code generated an error or not.